ledger/internal/eval.go

```diff
@@ -963,14 +970,34 @@ func (eval *BlockEvaluator) TransactionGroup(txgroup []transactions.SignedTxnWit
 	for gi, txad := range txgroup {
 		var txib transactions.SignedTxnInBlock
 
+		cowForTxn := cow
+
+		if eval.GranularEval {
+			cowForTxn = cow.child(1)
+		}
+
 		if eval.Tracer != nil {
 			eval.Tracer.BeforeTxn(evalParams, gi)
 		}
 
-		err := eval.transaction(txad.SignedTxn, evalParams, gi, txad.ApplyData, cow, &txib)
+		err := eval.transaction(txad.SignedTxn, evalParams, gi, txad.ApplyData, cowForTxn, &txib)
 
 		if eval.Tracer != nil {
-			eval.Tracer.AfterTxn(evalParams, gi, txib.ApplyData, err)
+			var update *ledgercore.StateDelta
+			if eval.GranularEval {
+				// Only include if we're sure the cowForTxn contains ONLY the updates from this txn
+				u := cowForTxn.Updates()
+				update = &u
+			}
+			eval.Tracer.AfterTxn(evalParams, gi, txib.ApplyData, update, err)
+		}
+
+		if eval.GranularEval {
+			cowForTxn.commitToParent()
```

Comment:
you can implement `sharedChild` for retrieving a new child and a new non-destructive `commitToParent`. Either try it or remove the comment below. I do not see any reason to do it since `recycle` + `child` efficiently give you the same functionality

Reply:
Yes this is a good idea

Reply:
but I do not see any reason to do it. The `GranularEval` is debugging feature and should not be running on production node.

Reply:
This code will likely run in a production node to implement the `/v2/transactions/simulate` endpoint, and it may be used independently by conduit at some point, so I think this optimization is worthwhile to consider.

I also worry that frequently touching the underlying `sync.Pool` like it does currently would have negative effects on other threads using the same pool. But this is less substantiated.

Reply:
I made this change in c07ccd245. Unfortunately it's not as easy to do the same thing for inner transactions, which I note here: https://github.com/algorand/go-algorand/blob/c07ccd24571a2a8819895d0dd4eb202f7af9df4c/ledger/eval/applications.go#L299-L301